### PR TITLE
[SPARK-11374][SQL] Support `skip.header.line.count` option for Hive Table

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
## What changes were proposed in this pull request?

This PR adds `skip.header.line.count` option support for Hive tables (external and managed).

- `CREATE EXTERNAL TABLE` or `CREATE TABLE ... LOCATION` uses external files as a data source for the tables. External files like CSV frequently contains one or more header lines as their own metadata. So, **skip.header.line.count** table property is used in order to exclude those metadata lines from data lines.
- Managed tables also have the same situations to handle files loaded by `LOAD DATA` SQL commands. 

[SPARK-11374](https://issues.apache.org/jira/browse/SPARK-11374) is a long lasting issue since 1.5.1, and all current Spark versions have the same situation.

The following is an example for SPARK-11374.

**CSV Data**

``` bash
$ cat /data/test.csv 
c1,c2
1,a
2,b
```

**Scala**

``` scala
scala> spark.read.option("header","true").csv("/data").show
+---+---+
| c1| c2|
+---+---+
|  1|  a|
|  2|  b|
+---+---+
```

**SQL (External Table)**

``` scala
scala> sql("CREATE TABLE t1 (id INT, value VARCHAR(10)) ROW FORMAT DELIMITED FIELDS TERMINATED BY ',' STORED AS TEXTFILE LOCATION '/data' TBLPROPERTIES('skip.header.line.count'='1')")
scala> sql("SELECT * FROM t1").show
+---+-----+
| id|value|
+---+-----+
|  1|    a|
|  2|    b|
+---+-----+
```

**SQL (Managed Table)**

```scala
scala> sql("CREATE TABLE t2 (id INT, value VARCHAR(10)) ROW FORMAT DELIMITED FIELDS TERMINATED BY ',' TBLPROPERTIES('skip.header.line.count'='1')")
scala> sql("LOAD DATA LOCAL INPATH '/data/test.csv' OVERWRITE INTO TABLE t2")
scala> sql("SELECT * FROM t2").show
+---+-----+
| id|value|
+---+-----+
|  1|    a|
|  2|    b|
+---+-----+
```

## How was this patch tested?

Pass the Jenkins test with a new test case.

### CODE REVIEW

**Comment by rxin** [comment]
Is this a CSV specific issue? If yes, it seems wrong to have logic implemented for all hive serdes for this thing.

**Comment by dongjoon-hyun** [comment]
Actually, I described like that for simplicity. But, CSV is one of the typical case of this option. This issue is not a CSV specific issue. I think this is a general table properties as you see the following.
```
CREATE TABLE t1 (id INT, b VARCHAR(10))
ROW FORMAT DELIMITED
FIELDS TERMINATED BY ','
STORED AS TEXTFILE LOCATION '/data'
TBLPROPERTIES('skip.header.line.count'='1')
```

**Comment by dongjoon-hyun** [comment]
Hi, @rxin . I update the PR description first.

**Comment by rxin** [comment]
Does this option make sense in Parquet or ORC?

**Comment by dongjoon-hyun** [comment]
I see. Thank you for guidance. Initially, I thought this table property is orthogonally skipping the starting rows of each file for all formats. And, I assumed that users will not use this property improperly.

But, what you mean is it has no meaning for columnar and vectorized formats. So, if a user give this table property for this Parquet or ORC, Spark need to ignore this.

If then, definitely, we should find some places for TEXT format only. BTW, do you have some proper location in your mind instead of the current `hadoopRDD.mapPartitionsWithIndex`?

**Comment by dongjoon-hyun** [comment]
Hi, @rxin . Until now, I couldn't find more suitable class in `hive` package. I added the logic to check the Table's InputFormat. Now, `skip.header.line.count` option is applied for the table using `org.apache.hadoop.mapred.TextInputFormat` only.

**Comment by dongjoon-hyun** [comment]
Rebased to the master.

**Comment by jamartinh** [comment]
This is necessary, please add this to a release !

**Comment by dongjoon-hyun** [comment]
Thank you, @jamartinh . I hope so.

**Comment by dongjoon-hyun** [comment]
Retest this please.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala
Comment (srowen) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala]: Instead of this, use `.toInt` to parse as an int in Scala?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala
Comment (srowen) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala]: Again just for Scala-like syntax, use == for strings?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala
Comment (srowen) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala]: `else break` or something? or else these keeps querying the iterator. Maybe a while loop would even be nice but I'm not sure.

**Review by srowen — COMMENTED** [review]
Ah, I see this has been around for a while. We should decide what to do. I added a few style changes. It does seem reasonable but I think the question here is whether the change belongs here or elsewhere. @rxin if you have a moment, does this look convincing or did you still have a concern about it?

**Comment by dongjoon-hyun** [comment]
Thank you for review, @srowen !
Yes. The issue was `the question here is whether the change belongs here or elsewhere`, and I couldn't find better place.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala
Comment (dongjoon-hyun) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala]: Sure.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala
Comment (dongjoon-hyun) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala]: Correct. I will change to use `while`.

**Comment by dongjoon-hyun** [comment]
Retest this please

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala
Comment (srowen) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala]: Can this be something like `hiveTable.getInputFormatClass.isInstanceOf[TextInputFormat]`?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala
Comment (srowen) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala]: I wonder if it's best to be defensive and test whether these things are actually instances of the classes they're expected to be? or is that too paranoid?

**Review by srowen — CHANGES_REQUESTED** [review]
@liancheng or @yhuai do you have thoughts on this? it seems fairly reasonable to me.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala
Comment (dongjoon-hyun) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala]: Sure. I'll fix like that.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala
Comment (yhuai) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala]: is `partition.inputSplit.t.asInstanceOf[FileSplit].getStart() != 0` tested (for a split that does not start from the start of the file, we do not skip)?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala
Comment (yhuai) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala]: I am +1 on adding the check.

Comment (dongjoon-hyun) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala]: Oh, thank you for review, @yhuai ! Yep.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala
Comment (dongjoon-hyun) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala]: Yep. I'll add that, too.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala
Comment (srowen) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala]: Oh right, I guess I mean ... `hiveTable.getInputFormatClass == classOf[TextInputFormat]`?

Comment (dongjoon-hyun) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala]: Oh, right. I see.

**Review by srowen — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
Thank you for approving, @srowen !

**Comment by rxin** [comment]
@dongjoon-hyun  

can you explain why this is only a problem for external table but not managed tables?

**Comment by dongjoon-hyun** [comment]
@rxin .
I didn't say this is only a problem for external table.
I simply focused on SPARK-11374 use case here.

What use case are you considering when you mentioned **managed tables**?

**Comment by rxin** [comment]
The way the description is written implies that there is something special about external tables or tables with location specified. I'm just confused here.

**Comment by dongjoon-hyun** [comment]
Ah, I see.
Then, I'll try to update PR descriptions or testcases more.

**Comment by dongjoon-hyun** [comment]
@rxin .
I updated the title and description of this PR.
For the test case, I think the current test case covers enough the functionality.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/SQLQuerySuite.scala]: Oh, I missed to remove some tables. I'll update the testcode.

**Comment by rxin** [comment]
@dongjoon-hyun can you show a link to how Hive implements this? I'm just surprised that this is a general feature rather than something that is format specific.

**Comment by dongjoon-hyun** [comment]
I didn't read Hive code part for this. I referenced the Spark csv part code.
But, I'll search the corresponding one in Hive master branch tomorrow since you requested.

**Comment by dongjoon-hyun** [comment]
Hi, @rxin .
Here is the corresponding HIve implementation.

**getNextRow**

* https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/exec/FetchOperator.java#L470-L474

```
          headerCount = Utilities.getHeaderCount(currDesc.getTableDesc());
          footerCount = Utilities.getFooterCount(currDesc.getTableDesc(), job);

          // Skip header lines.
          opNotEOF = Utilities.skipHeader(currRecReader, headerCount, key, value);
```

**skipHeader**

* https://github.com/apache/hive/blob/master/ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java#L3385-L3393

```
  public static boolean skipHeader(RecordReader<WritableComparable, Writable> currRecReader,
      int headerCount, WritableComparable key, Writable value) throws IOException {
    while (headerCount > 0) {
      if (!currRecReader.next(key, value))
        return false;
      headerCount--;
    }
    return true;
  }
```

**Comment by dongjoon-hyun** [comment]
@rxin 
Could you give me any advice for this issue?

**Comment by dongjoon-hyun** [comment]
Hi, @rxin .
Do you want me to prepare something based on `InputFileBlockHolder` of #16133 ?

**Comment by dongjoon-hyun** [comment]
Could you review this again please when you have sometime, @rxin ?

**Comment by dongjoon-hyun** [comment]
Hi, @rxin , @srowen , @yhuai .
If there is something for me to do more for this issue, please let me know.

**Comment by dongjoon-hyun** [comment]
Retest this please

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala
Comment (mridulm) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala]: What about subclasses ? isAssignableFrom might be better ?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala
Comment (mridulm) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala]: There is an assumption here that 'skipHeaderLineCount' lines will be in the first split always ?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala
Comment (dongjoon-hyun) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala]: Sure! I'll update like that.

**Comment by dongjoon-hyun** [comment]
Thank you for review, @mridulm .
I updated the PR by using `isAssignableFrom`.

**Comment by dongjoon-hyun** [comment]
Retest this please

**Comment by gatorsmile** [comment]
In the original Hive JIRA of this feature: https://issues.apache.org/jira/browse/HIVE-5795 

See the latest reply from `Sergey Shelukhin`. 
> This forces the entire input into a single split, which defeats the purpose of using Hive in the first place - might as well run the analysis on a local machine. I would not recommend anyone to use this feature except for experimentation. The headers/footers should be cleared as part of an ETL process.

I do not know how many Hive users are using them in the production system. This might not be recommended in the Hive side.

**Comment by dongjoon-hyun** [comment]
Hi, @gatorsmile .
Do you think it's valid comment for this implementation? Why **'this forces entire input into a single split'**? We are not doing such a thing.
For me, it looks like a comment for `footer`. We didn't provide `footer`.

Also, the approach in this PR is based on Spark `spark.read.option("header","true").csv("/data").show` behavior, isn't it?

**Comment by gatorsmile** [comment]
If Hive users are not using this feature, should we still use the table property as the interface to do it?

Asking users to change the table property looks hacky to me.

**Comment by dongjoon-hyun** [comment]
Why do you think `Hive user are not using this feature`?

**Comment by gatorsmile** [comment]
That is my question. I do not know the answer. Based on the comment in the Hive JIRA, I am doubting how many Hive users are using it.

**Comment by dongjoon-hyun** [comment]
Ah, if that is the question, I knew the customers using this and requesting this. :)

**Comment by dongjoon-hyun** [comment]
Actually, it's a Spark issue SPARK-11374, isn't it?
So, the remaining question should be `the Spark user do want this`.
For this, I'll send an email to user and dev mailing list.
It's simple.

**Comment by gatorsmile** [comment]
In the Hive document, this is not included. https://cwiki.apache.org/confluence/display/Hive/LanguageManual+DDL#LanguageManualDDL-CreateTable

The list does not include it:
```
TBLPROPERTIES ("comment"="table_comment")
TBLPROPERTIES ("hbase.table.name"="table_name") – see HBase Integration.
TBLPROPERTIES ("immutable"="true") or ("immutable"="false") in release 0.13.0+ (HIVE-6406) – see Inserting Data into Hive Tables from Queries.
TBLPROPERTIES ("orc.compress"="ZLIB") or ("orc.compress"="SNAPPY") or ("orc.compress"="NONE") and other ORC properties – see ORC Files.
TBLPROPERTIES ("transactional"="true") or ("transactional"="false") in release 0.14.0+, the default is "false" – see Hive Transactions.
TBLPROPERTIES ("NO_AUTO_COMPACTION"="true") or ("NO_AUTO_COMPACTION"="false"), the default is "false" – see Hive Transactions.
TBLPROPERTIES ("compactor.mapreduce.map.memory.mb"="mapper_memory") – see Hive Transactions.
TBLPROPERTIES ("compactorthreshold.hive.compactor.delta.num.threshold"="threshold_num") – see Hive Transactions.
TBLPROPERTIES ("compactorthreshold.hive.compactor.delta.pct.threshold"="threshold_pct") – see Hive Transactions.
TBLPROPERTIES ("auto.purge"="true") or ("auto.purge"="false") in release 1.2.0+ (HIVE-9118) – see Drop Table and Drop Partitions.
TBLPROPERTIES ("EXTERNAL"="TRUE") in release 0.6.0+ (HIVE-1329) – Change a managed table to an external table and vice versa for "FALSE".
```

More importantly, we have not officially allowed users to use TBLPROPERTIES in Spark. I do not know whether this is the right way to do it. cc @rxin

**Comment by gatorsmile** [comment]
In DDL for creating data source tables, `OPTIONS` is being used for table-level parameters. In the near future, we need to combine DDL for data source tables and Hive serde tables. I am not sure whether `OPTIONS` is a better place for this purpose.

**Comment by gatorsmile** [comment]
Also cc @cloud-fan @hvanhovell

**Comment by dongjoon-hyun** [comment]
Thank you for the investigation. Is the Hive table property list exhaustive?

**Comment by gatorsmile** [comment]
That list is not exhaustive, but should we follow what we did for CSV, instead of using this hidden Hive feature? Should we use `OPTIONS` instead of `TBLPROPERTIES`?

**Comment by dongjoon-hyun** [comment]
Yes. Right, that is possible, but it's a different use case.

This PR supports the followings:
- The **existing Hive table** with `TBLPROPERTIES(skip.header.line.count=1)` will show the same result in Hive & Spark.
  Previously, Spark ignores that option. So, Spark shows incorrect result for that table.
- And, vice versa, the **table created by Spark** will show the same result in Hive & Spark.

This is not a hidden Hive feature. It was
- Reported by `Daniel Haviv` (Oct. 2015)
- Asked by `Stephane Maarek` (Apr. 2016)
- Asked by `Rahul Jain` (Sep. 2016)

Also, there is a StackOverflow question.
- http://stackoverflow.com/questions/38566591/spark-sql-hivecontext-dont-ignore-header

**Comment by cloud-fan** [comment]
I think this is really an uncommon hive feature, and users can work around it by creating an external data source table on the CSV files.

And another concern is, we can't simply call some Hive APIs, but have to reimplement this feature at Spark side(in the hive table reader). If there are more Hive features like this, we will complicate the hive table reader quite a bit.

**Comment by HyukjinKwon** [comment]
I know votes and stars might not be an appropriate standard for the popularity of a feature but I found another stackoverflow question - http://stackoverflow.com/questions/15751999/hive-external-table-skip-first-row

**Comment by cloud-fan** [comment]
@HyukjinKwon yea for Hive this feature is useful, as the only workaround is running a ETL program to remove the headers manually. However, in Spark SQL, our CSV data source already support skipping header lines, the problem is external hive serde table, and users can easily work around it by creating an external data source table on the CSV files.

**Comment by dongjoon-hyun** [comment]
Thank you for opinion, @HyukjinKwon and @cloud-fan. 

BTW, our CSV data sources does not support multiple line headers so far.

**Comment by dongjoon-hyun** [comment]
Now, I think we have been considered almost pros and cons here. Thank you all for reviewing and giving opinions for this issue.

I value the committers' decision and feel the worry from @rxin , @gatorsmile , @cloud-fan .

The overall opinion seems to be it's not a proper Spark way for Spark users, right?

**Comment by jamartinh** [comment]
My humble comment is simply that an external Hive table still is a Hive
Table, and as such it should be transparent for whichever user how to work
with it, even for spark users like me. The expectation is that a query from
such table yield the same results as if it where queried from Hive. I have
seen people worried about this.

That is, it is very unexpected querying a table from Hive and obtain
different result when queried from Spark.

I still consider this functionality as a neccesary feature in order to
maintain consistency with Hive engine, otherwise spark hive datasource
 will slowly be diverging from Hive. I feel this should not be the case.

Best,
Jose



El 9 dic. 2016 23:42, "Dongjoon Hyun" <notifications@github.com> escribió:

Now, I think we have been considered almost pros and cons here. Thank you
all for reviewing and giving opinions for this issue.

I value the committers' decision and feel the worry from @rxin
<https://github.com/rxin> , @gatorsmile <https://github.com/gatorsmile> ,
@cloud-fan <https://github.com/cloud-fan> .

The overall opinion seems to be it's not a proper Spark way for Spark
users, right?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/apache/spark/pull/14638#issuecomment-266143334>, or mute
the thread

**Comment by srowen** [comment]
I am really on the fence about it. I can see the argument that it's useful to support, and not hard to support. But also that this is just one of many Hive features that one could support, and we have to support it separately, and it duplicates other Spark-specific mechanisms. 

For what it's worth we did hear from a few users who said this would be useful. If we don't merge this, then we're saying that we believe there's a different and better way to do this via Spark CSV support. Is that true -- do we think that's not a big inconvenience?

**Comment by HyukjinKwon** [comment]
Up to my knowledge, currently, CSV datasource does supports to skip a (only) single line with `header` option and also `comment` option to skip the lines started with the character given to `comment` whereas `skip.header.line.count` suggested here literally supports to skip multiple lines. 

So, strictly they are not exactly duplicates but I believe it is true some ranges that `skip.header.line.count` supports can be worked around with the options, `header` and `comment` as explained by @dongjoon-hyun and @cloud-fan.

If we admit skipping a single line (or skipping lines started with `comment`) is a common case and enough to cover this feature, that would not be a big inconvenience but if users believe it is still useful to skip multiple lines, we would still need this.

**Comment by gatorsmile** [comment]
@jamartinh I got your point. Hive tables can be read/wrote by both Hive and Spark. Ideally, we should respect all the features supported by Hive. That means, if the metadata of an existing Hive table already has such a `TBLPROPERTIES`, we should not simply ignore it. 

However, when users want to add it in Spark side, what is the right interface? In the existing Spark SQL, table properties are not used for such purposes. When users specify the table-level properties/options for data source tables, we store them in the storage property. Thus, they will not be recognized by Hive. We might need to hold it until we figure out how to unify the SQL interfaces of Hive tables and the other data source tables. `TBLPROPERTIES` is not the right interface for us, at least now.

**Comment by dongjoon-hyun** [comment]
Thank you so much, @jamartinh , @srowen , @HyukjinKwon , and @gatorsmile .

We can distinguish the two existing problems separately here.

First, **a)** Spark returns incorrect result for an existing Hive table already with `skip.header.line.count` table property. This is the most common use case which this issue aimed to solve.

Second, more ridiculously, **b)** Spark can create a table with `skip.header.line.count` table property and only Hive returns the correct result from that table.

**SPARK (Current master branch)**
```scala
scala> sql("CREATE TABLE t2 (id INT, value VARCHAR(10)) ROW FORMAT DELIMITED FIELDS TERMINATED BY ',' TBLPROPERTIES('skip.header.line.count'='1')")

scala> sql("LOAD DATA LOCAL INPATH '/data/test.csv' OVERWRITE INTO TABLE t2")

scala> sql("SELECT * FROM t2").show
+----+-----+
|  id|value|
+----+-----+
|null|   c2|
|   1|    a|
|   2|    b|
+----+-----+
```
**Hive**
```scala
hive> select * from t2;
OK
1	a
2	b
```

@gatorsmile . I totally agree on the Apache Spark development direction.  But, IMO, `TBLPROPERTIES` or `OPTION` is not a proper issue in this PR. It's because this PR only updates `TableReader.scala` to support the existing table property, case **a)**. For `TBLPROPERTIES`, I simply used that because it's already supported on Spark. I can update the PR description in order to focus on **a)** instead of **b)**.

Someday later, Apache Spark may delete(or block) `TBLPROPERTIES` SQL syntax in favor of `OPTION` syntax. It's okay. It's just a kind of regression on purpose. No problem at all. However, even in that case, we had better read the Hive table with `skip.header.line.count` correctly.

**Comment by cloud-fan** [comment]
Is there any high-level API that we can use to read hive tables? The current hive table reader is so low-level that we have to support features like `skip.header.line.count`, and I think it doesn't work well with non-file based hive tables.

**Comment by dongjoon-hyun** [comment]
Do you mean to replace the current whole `TableReader.scala` which was introduced in SPARK-1251 ? I guessed Spark chose this direct access approach for the performance issue at that time.

Yes. This option is targeting only `TextInputFormat`s. For non-file based hive tables like Orc/Parquet, this option is ignored.
```scala
val isTextInputFormatTable = classOf[TextInputFormat].isAssignableFrom(hiveTable.getInputFormatClass)
```

**Comment by dongjoon-hyun** [comment]
Thank you all so much for reviewing and discussion on this issue since Aug 14th.
Since we focused this fully both here and in Spark mailing list and it has been 4 months already, I'll happily close this PR and SPARK-11374 as a WON'T FIX tomorrow (Dec. 14th).
All of your time is too precious to be stuck here any more. Apache Spark still have many things to do. Go Spark!

**Comment by Derek-Lin** [comment]
Sadly, it's a useful feature which i want.
i want to retreive HDFS tsv files via spark sql jdbc interface which defined as hive table

**Comment by sadhen** [comment]
Also a useful feature for me.

**Comment by sadhen** [comment]
Since this pr won't be accepted. My solution is:
``` sql
# create table using spark-csv
CREATE TABLE tmp.cars
USING com.databricks.spark.csv
OPTIONS (path "/tmp/cars.csv", header "true", inferSchema "true");

# create table use CTAS: whether using parquet or orc or others depends on your conf
create table tmp.cars_transformed as select * from tmp.cars

# optional: drop the spark-csv table
drop table tmp.cars
```

**Comment by gatorsmile** [comment]
@sadhen Your solution works well without any code change, right?

**Comment by sadhen** [comment]
yes

**Comment by set92** [comment]
@sadhen Using pySpark 2.4.0 and I still have this issue, in the repo of spark.csv it says "This functionality has been inlined in Apache Spark 2.x." so that means that tblproperties should work? or I have to modify something in your code to make it work in the new versions?

**Comment by dongjoon-hyun** [comment]
Nope, please don't expect that, @set92 . BTW, where did you see the message? 
> This functionality has been inlined in Apache Spark 2.x

**Comment by set92** [comment]
@dongjoon-hyun in the [README of spark-csv](https://github.com/databricks/spark-csv), which I read was one option to solve this problem pre2.0.

**Comment by dongjoon-hyun** [comment]
I'm not sure why you ask that question here. This is Apache Spark repo and Apache Spark community didn't write that sentence. Could you ask there, @set92 ? :)

**Comment by srowen** [comment]
It's a fair question as it was created by some of the same people as here, and we advertised it as merged into Spark. I don't know if 100% of the functionality was ported as is. But it's a question for user@, really.

**Comment by dongjoon-hyun** [comment]
Got it, @srowen .

**Comment by asolimando** [comment]
We are having the same issue on our side as well, this patch would have been terribly useful to us.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala
Comment (liujiayi771) [reply sql/hive/src/main/scala/org/apache/spark/sql/hive/TableReader.scala:142]: @dongjoon-hyun I would like to use this PR, but it seems that in the current Spark version, performing an RDD transformation within another RDD transformation is not allowed, which results in an `RDDLacksSparkContextError()`. Do you know how to modify this?

**Pull request closed without merging**
